Seed SecureRandom with -Dtests.seed for JWK RSA/EC/HMAC keygen in JWT tests

justincr-elastic opened the pull request:
The purpose of this PR is to support easy reproduction of an [intermittent CI Test failure](https://github.com/elastic/elasticsearch/issues/89509).

I strongly suspect it is hard to reproduced because JWK keygen is using default SecureRandom, and there is no LuceneTestCase.secureRandom() available. LuceneTestCase only offers random() for using -Dtests.seed.

This PR depends on [PR #93](https://bitbucket.org/connect2id/nimbus-jose-jwt/pull-requests/93/add-support-to-use-securerandom-in) submitted to nimbus-jose-jwt. My PR was accepted into [v9.25](https://mvnrepository.com/artifact/com.nimbusds/nimbus-jose-jwt/9.25).

https://bitbucket.org/connect2id/nimbus-jose-jwt/src/master/CHANGELOG.txt
```
version 9.25 (2022-09-13)
    * Adds support to use a SecureRandom in RSAKeyGenerator.generate() and
      ECKeyGenerator.generate().
```

justincr-elastic commented:
Adding @mark-vieira as a reviewer since this is related to test reproduction infrastructure.

When I run with the current code changes, I am getting an access denied error from Security Manager. I am looking for feedback how to workaround this issue, either via a code change or a config change. I think it relates to however elasticsearch is using a custom Security Manager, and I am not familiar with those details.

```
REPRODUCE WITH: ./gradlew ':x-pack:plugin:security:test' --tests "org.elasticsearch.xpack.security.authc.jwt.JwtRealmAuthenticateTests.testJwkSetUpdates" -Dtests.seed=E930884F377FE425 -Dtests.locale=ar-SD -Dtests.timezone=Asia/Makassar -Druntime.java=18

access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
	at __randomizedtesting.SeedInfo.seed([E930884F377FE425:DBEF29C6CE70B5B1]:0)
	at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:485)
	at java.base/java.security.AccessController.checkPermission(AccessController.java:1068)
	at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:411)
	at java.base/java.lang.Class.checkMemberAccess(Class.java:3083)
	at java.base/java.lang.Class.getDeclaredConstructor(Class.java:2782)
	at com.nimbusds.jose.shaded.gson.internal.ConstructorConstructor.newDefaultConstructor(ConstructorConstructor.java:193)
	at com.nimbusds.jose.shaded.gson.internal.ConstructorConstructor.get(ConstructorConstructor.java:101)
	at com.nimbusds.jose.shaded.gson.internal.bind.MapTypeAdapterFactory.create(MapTypeAdapterFactory.java:126)
	at com.nimbusds.jose.shaded.gson.Gson.getAdapter(Gson.java:531)
	at com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:778)
	at com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:756)
	at com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:711)
	at com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:691)
	at com.nimbusds.jose.util.JSONObjectUtils.toJSONString(JSONObjectUtils.java:523)
	at org.elasticsearch.xpack.security.authc.jwt.JwtUtil.serializeJwkSet(JwtUtil.java:228)
	at org.elasticsearch.xpack.security.authc.jwt.JwtIssuer.setJwks(JwtIssuer.java:90)
	at org.elasticsearch.xpack.security.authc.jwt.JwtRealmTestCase.createJwtIssuer(JwtRealmTestCase.java:216)
	at org.elasticsearch.xpack.security.authc.jwt.JwtRealmTestCase.generateJwtIssuerRealmPairs(JwtRealmTestCase.java:163)
	at org.elasticsearch.xpack.security.authc.jwt.JwtRealmAuthenticateTests.testJwkSetUpdates(JwtRealmAuthenticateTests.java:102)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
	at java.base/java.lang.Thread.run(Thread.java:833)
111

justincr-elastic commented:
@elasticmachine update branch

justincr-elastic commented:
@elasticmachine update branch

mark-vieira commented:
@ChrisHegarty might have some ideas on how to get around the security manager problem, although if this PR is never meant to be merged and we're just trying to repo, you could just temporarily add the necessary permission [here](https://github.com/elastic/elasticsearch/blob/600cb9108071a63f2c63a8fffa21c34e15251433/server/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy).

justincr-elastic commented:
I clobbered the PR contents when copying a temp fix over from a different PR. Trying to restore.

justincr-elastic committed:
Use -Dtests.seed for SecureRandom

justincr-elastic committed:
Copy docker-compose workaround from #90457

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into citest/random-failures-89509-use-securerandom-with-seed

justincr-elastic commented:
> @ChrisHegarty might have some ideas on how to get around the security manager problem, although if this PR is never meant to be merged and we're just trying to repro a failure, you could just temporarily add the necessary permission [here](https://github.com/elastic/elasticsearch/blob/600cb9108071a63f2c63a8fffa21c34e15251433/server/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy).

The library update and test changes are meant to be permanent.

ChrisHegarty commented:
It's curious why this permission is now suddenly an issue, i.e. what changed to cause the permission to fail?

To create some test specific data, the test is using `com.nimbusds.jose.util.JSONObjectUtils`, which internally uses `Gson` that ends up invoking `Class::getDeclaredConstructor` - which ultimately triggers the permission check. Why was Gson working ok before? There's likely a lot of changes between `nimbus-jose-jwt:9.8.1` and `nimbus-jose-jwt:9.25`, so maybe this was just introduced by some internal refactoring. If so, then adding the permission to the test framework permission grant is the right permanent solution.

justincr-elastic commented:
@ChrisHegarty, there are two changes.
1. Add calls to `SecureRandom.getStrongInstance()` and `SecureRandom.setSeed(LuceneTestCase.randomBytes(32))`.
2. Pass SecureRandom instance to updated RSA/EC JwkGenerator utils inside nimbus-jose-jwt:9.25.

The security management exception is triggers by 1. The call to the updated library has not taken place yet.

The original problem is LuceneTestCase has random() but not secureRandom(). Short term, I am trying to use random() to seed a SecureRandom, to make it deterministic based on -Dtests.seed. Long term, it would be nice to move this into a new LuceneTestCase.secureRandom() so it is available for all test code to use.

justincr-elastic committed:
Copy docker-compose workaround from #90457

justincr-elastic committed:
Use getInstance instead of getInstanceStrong

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into citest/random-failures-89509-use-securerandom-with-seed

justincr-elastic committed:
More DockerSupport logging

justincr-elastic committed:
Logging and checkStyle

justincr-elastic committed:
Ignore missing nimbus dep class XChaCha20Poly1305

justincr-elastic committed:
Update jars

justincr-elastic committed:
Try to update policy

justincr-elastic commented:
@ChrisHegarty @mark-vieira I tried to add `accessDeclaredMembers` to policy in [050e624](https://github.com/elastic/elasticsearch/pull/90315/commits/050e62426d66656926b2fbc00cae807b5ead0da0), but it is not working. I am not familiar with how to update the policy, so I don't know if I am doing it right.

justincr-elastic committed:
Remove `codeBase "${codebase.nimbus-jose-jwt}"`

justincr-elastic committed:
Remove `codeBase "${codebase.nimbus-jose-jwt}"`

justincr-elastic committed:
Cleanup 3rd party audit missing classes task

justincr-elastic committed:
Add another permission

justincr-elastic committed:
Use default SecureRandom constructor to fix FIPS

justincr-elastic committed:
Temp override to log seed

justincr-elastic commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

justincr-elastic committed:
Merge branch 'main' into citest/random-failures-89509-use-securerandom-with-seed

justincr-elastic commented:
@elasticmachine update branch

elasticmachine commented:
expected head sha didn’t match current head ref.

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into citest/random-failures-89509-use-securerandom-with-seed

justincr-elastic committed:
Remove temp Docker Compose workaround

elasticmachine committed:
Merge branch 'main' into citest/random-failures-89509-use-securerandom-with-seed

ChrisHegarty commented on server/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy:
```diff
@@ -125,4 +125,10 @@ grant codeBase "${codebase.netty-transport}" {
+  permission java.lang.RuntimePermission "accessDeclaredMembers";
+
+  // needed for com.nimbusds.oauth2.sdk.id.Identifier#value
+  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
```
Can we grant these permissions be granted to the nimbusds codebase, rather than the "global" grant?

justincr-elastic replied:
I agree it would be better to scope it to the two jars. However, I tried this and it didn't work.

```
grant codeBase "${codebase.nimbus-jose-jwt}" {
  // needed for com.nimbusds.jose.shaded.gson.Gson.toJson
  permission java.lang.RuntimePermission "accessDeclaredMembers";
};

grant codeBase "${codebase.oauth2-oidc-sdk}" {
  // needed for com.nimbusds.oauth2.sdk.id.Identifier#value
  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
};
```

I tried adding `nimbus-jose-jwt` and `oauth2-oidc-sdk` entries here, but it didn't work.
https://github.com/elastic/elasticsearch/blob/64728c08422ce2d493502809bb09979a4aa6852b/test/framework/src/main/java/org/elasticsearch/bootstrap/BootstrapForTesting.java#L212-L224

There is a check here which rejects classes in third-party jar files, so it was not clear to me how to add the two codebases without interfering with the existing, internal codebases.
https://github.com/elastic/elasticsearch/blob/64728c08422ce2d493502809bb09979a4aa6852b/test/framework/src/main/java/org/elasticsearch/bootstrap/BootstrapForTesting.java#L227-L243

justincr-elastic replied:
This is what I tried to do:
```
addClassCodebase(codebases, "nimbus-jose-jwt", "com.nimbusds.jose.shaded.gson.Gson.toJson");
addClassCodebase(codebases, "oauth2-oidc-sdk", "com.nimbusds.oauth2.sdk.id.Identifier");
```

The `location` value in `addClassCodebase` ends with `.jar`, so this code falls through without registering the codebase.
```
if (location.toString().endsWith(".jar") == false) {
```

References to `codebase.nimbus-jose-jwt` and `codebase.oauth2-oidc-sdk` inside the policy files are unresolved, so the policy grant cannot be parsed correctly.

justincr-elastic replied:
@ChrisHegarty, I tried moving the policy changes from `server/src/main/resources/org/elasticsearch/bootstrap/security.policy` to `x-pack/plugin/security/src/main/plugin-metadata/plugin-security.policy`, but this part of the build failed: `./gradlew :x-pack:qa:oidc-op-tests:javaRestTest --debug --stacktrace`

Success (server): https://gradle-enterprise.elastic.co/s/mr3onfehxffj2/console-log#L886
Failure (x-pack security): https://gradle-enterprise.elastic.co/s/qdi5gmcqzq65i/console-log#L748

Does the same permission need to be added elsewhere? I looked under /x-pack/security/qa/, but I don't see a policy file in there.

justincr-elastic replied:
I debugged PolicyUtil and fo0und the grant names for these jars include the version and jar, so I updated the grants. I double checked and those policies work in unit tests. However, they don't help inside the elasticsearch:test container of the test fixture. In other words, SecurityManager policy seems to be treated differently inside vs outside the container.

ChrisHegarty commented on server/src/main/resources/org/elasticsearch/bootstrap/security.policy:
```diff
@@ -172,4 +172,9 @@ grant {
+  permission java.lang.RuntimePermission "accessDeclaredMembers";
+
+  // needed for com.nimbusds.oauth2.sdk.id.Identifier#value
+  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
```
Can we grant these permissions be granted to the nimbusds codebase, rather than the "global" grant?

justincr-elastic replied:
Duplicate of a different comment. Addressing it in the other comment.

elasticmachine committed:
Merge branch 'main' into citest/random-failures-89509-use-securerandom-with-seed

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

justincr-elastic committed:
Move policy change from server to x-pack security

justincr-elastic committed:
Add TODO to remove log message

justincr-elastic committed:
Attempt to add policy to x-pack:qa:oidc-op-tests

justincr-elastic committed:
Consolidate grants in x-pack policy

ChrisHegarty commented on x-pack/plugin/security/src/main/plugin-metadata/plugin-security.policy:
```diff
@@ -45,3 +45,11 @@ grant codeBase "${codebase.netty-transport}" {
+
+  // needed for com.nimbusds.oauth2.sdk.id.Identifier#value
+  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
+};
```
It's more straightforward to read and understand policy files if there is just a **single** grant to each codebase (or non-codebase/global).

justincr-elastic replied:
Agreed. I am blocked trying to get this to work in the oidc-op-tests test fixture, so I thought maybe the codebase names were not being read. If I can get past whatever is blocking oidc-op-tests from applying a new policy, I can separate the grants by codebase again.

justincr-elastic replied:
I separated them by codebase again. I tried with and without codebase, no luck. The stacktrace is in the latest PR comment.

justincr-elastic committed:
Try policy change again

justincr-elastic commented:
Still not having any luck making the `idp-fixture` work when running `./gradlew :x-pack:qa:oidc-op-tests:javaRestTest`
 
Here are the two new jar versions inside the `elasticsearch:test` container for the `idp-fixture`.
```
elasticsearch@a5e916359ac3:~$ ls -l /usr/share/elasticsearch/modules/x-pack-security/nimbus-jose-jwt-9.25.4.jar /usr/share/elasticsearch/modules/x-pack-security/oauth2-oidc-sdk-9.43.1.jar
-r--r--r-- 1 root root 681930 Oct  6 18:51 /usr/share/elasticsearch/modules/x-pack-security/nimbus-jose-jwt-9.25.4.jar
-r--r--r-- 1 root root 820689 Oct  6 18:51 /usr/share/elasticsearch/modules/x-pack-security/oauth2-oidc-sdk-9.43.1.jar
```
Here is the x-pack `plugin-security.policy` inside the `idp-container` with my new additions.
```
elasticsearch@de8b37bf3e0a:~$ tail -11 /usr/share/elasticsearch/modules/x-pack-security/plugin-security.policy
grant codeBase "${codebase.nimbus-jose-jwt}" {
  // needed for com.nimbusds.jose.shaded.gson.Gson.toJson
  permission java.lang.RuntimePermission "accessDeclaredMembers";
};

grant codeBase "${codebase.oauth2-oidc-sdk}" {
  // needed for com.nimbusds.oauth2.sdk.id.Identifier#value
  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
};
```

Those two permissions were needed to fix OIDC unit tests that run inside IntelliJ. The same permissions are not working for OIDC integ tests inside the `idp-fixture` `elasticsearch:test` container. Here is the exception copied from `/usr/share/elasticsearch/logs/console.log`. It corresponds to the first permission, so I was expecting the OIDC integ test to work.
```
java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
	at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:485)
	at java.base/java.security.AccessController.checkPermission(AccessController.java:1068)
	at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:411)
	at java.base/java.lang.Class.checkMemberAccess(Class.java:3083)
	at java.base/java.lang.Class.getDeclaredConstructor(Class.java:2782)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.shaded.gson.internal.ConstructorConstructor.newDefaultConstructor(ConstructorConstructor.java:193)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.shaded.gson.internal.ConstructorConstructor.get(ConstructorConstructor.java:101)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.shaded.gson.internal.bind.MapTypeAdapterFactory.create(MapTypeAdapterFactory.java:126)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.shaded.gson.Gson.getAdapter(Gson.java:531)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:778)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:756)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:711)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:691)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jose.util.JSONObjectUtils.toJSONString(JSONObjectUtils.java:527)
	at com.nimbusds.jose.jwt@9.25.4/com.nimbusds.jwt.JWTClaimsSet.toString(JWTClaimsSet.java:882)
	at oauth2.oidc.sdk@9.43.1/com.nimbusds.oauth2.sdk.util.JSONObjectUtils.toJSONObject(JSONObjectUtils.java:904)
	at oauth2.oidc.sdk@9.43.1/com.nimbusds.openid.connect.sdk.claims.IDTokenClaimsSet.<init>(IDTokenClaimsSet.java:238)
	at oauth2.oidc.sdk@9.43.1/com.nimbusds.openid.connect.sdk.validators.IDTokenValidator.toIDTokenClaimsSet(IDTokenValidator.java:372)
	at oauth2.oidc.sdk@9.43.1/com.nimbusds.openid.connect.sdk.validators.IDTokenValidator.validate(IDTokenValidator.java:322)
	at oauth2.oidc.sdk@9.43.1/com.nimbusds.openid.connect.sdk.validators.IDTokenValidator.validate(IDTokenValidator.java:254)
	at org.elasticsearch.security@8.6.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectAuthenticator.getUserClaims(OpenIdConnectAuthenticator.java:267)
	at org.elasticsearch.security@8.6.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectAuthenticator.lambda$authenticate$0(OpenIdConnectAuthenticator.java:224)
	at org.elasticsearch.server@8.6.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.security@8.6.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectAuthenticator.handleTokenResponse(OpenIdConnectAuthenticator.java:670)
	at org.elasticsearch.security@8.6.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectAuthenticator$2.completed(OpenIdConnectAuthenticator.java:588)
	at org.elasticsearch.security@8.6.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectAuthenticator$2.completed(OpenIdConnectAuthenticator.java:585)
```

justincr-elastic commented:
I tried a global policy, no codebase, but that didn't work either.
```
elasticsearch@55d8e5e70d03:~/modules/x-pack-security$ tail -9 /usr/share/elasticsearch/modules/x-pack-security/plugin-security.policy
grant {
  // needed for com.nimbusds.jose.shaded.gson.Gson.toJson
  permission java.lang.RuntimePermission "accessDeclaredMembers";

  // needed for com.nimbusds.oauth2.sdk.id.Identifier#value
  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
};
```

justincr-elastic commented:
I added this to the start of Elasticsearch.main.
```
    public static void main(final String[] args) {
        System.out.println("JUSTIN Elasticsearch.main JarHell.parseModulesAndClassPath");
        for (URL url : JarHell.parseModulesAndClassPath()) {
            try {
                System.out.println("JUSTIN " + PathUtils.get(url.toURI()));
            } catch (URISyntaxException e) {
                e.printStackTrace();
                throw new RuntimeException(e);
            }
        }
```
I don't see the `x-pack-security` plugin or its `plugin-security.policy` on this classpath.
```
elasticsearch@13f545e25a37:~$ grep JUSTIN /usr/share/elasticsearch/logs/console.log 
JUSTIN Elasticsearch.main JarHell.parseModulesAndClassPath
JUSTIN /usr/share/elasticsearch/lib/lucene-analysis-common-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-core-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-plugin-classloader-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/jna-5.10.0.jar
JUSTIN /usr/share/elasticsearch/lib/ecs-logging-core-1.2.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-secure-sm-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/log4j2-ecs-layout-1.2.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/jopt-simple-5.0.2.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-queryparser-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/HdrHistogram-2.1.9.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-sandbox-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-plugin-analysis-api-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/t-digest-3.2.jar
JUSTIN /usr/share/elasticsearch/lib/lz4-java-1.8.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-cli-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-grouping-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/log4j-api-2.19.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-x-content-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-join-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-suggest-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-plugin-api-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-geo-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-backward-codecs-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-logging-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-log4j-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-queries-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/hppc-0.8.1.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-misc-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-highlighter-9.4.0.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-lz4-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/elasticsearch-core-8.6.0-SNAPSHOT.jar
JUSTIN /usr/share/elasticsearch/lib/lucene-memory-9.4.0.jar
```

How/where does a module like `x-pack-security` get loaded?

What I would like to do is add more logs like above to confirm what is being loaded from x-pack-security. I assume `oauth2-oidc-sdk-9.43.1.jar` and `nimbus-jose-jwt-9.25.4.jar` are loaded from the `x-pack-security` module directory, because their classes are listed in the original stacktrace. I would like to see if `plugin-security.policy` is possibly being ignored.
```
elasticsearch@13f545e25a37:~$ ls -l /usr/share/elasticsearch/modules/x-pack-security/
LICENSE.txt                                          metrics-core-4.1.4.jar                               opensaml-saml-api-4.0.1.jar
NOTICE.txt                                           netty-buffer-4.1.82.Final.jar                        opensaml-saml-impl-4.0.1.jar
accessors-smart-2.4.8.jar                            netty-codec-4.1.82.Final.jar                         opensaml-security-api-4.0.1.jar
asm-9.4.jar                                          netty-codec-http-4.1.82.Final.jar                    opensaml-security-impl-4.0.1.jar
cryptacular-1.2.4.jar                                netty-common-4.1.82.Final.jar                        opensaml-soap-api-4.0.1.jar
failureaccess-1.0.1.jar                              netty-handler-4.1.82.Final.jar                       opensaml-soap-impl-4.0.1.jar
guava-28.2-jre.jar                                   netty-resolver-4.1.82.Final.jar                      opensaml-storage-api-4.0.1.jar
httpclient-cache-4.5.13.jar                          netty-transport-4.1.82.Final.jar                     opensaml-storage-impl-4.0.1.jar
jakarta.mail-2.0.1.jar                               netty-transport-native-unix-common-4.1.82.Final.jar  opensaml-xmlsec-api-4.0.1.jar
java-support-8.0.0.jar                               nimbus-jose-jwt-9.25.4.jar                           opensaml-xmlsec-impl-4.0.1.jar
jcip-annotations-1.0.jar                             oauth2-oidc-sdk-9.43.1.jar                           plugin-descriptor.properties
joda-time-2.10.10.jar                                opensaml-core-4.0.1.jar                              plugin-security.policy
json-smart-2.4.8.jar                                 opensaml-messaging-api-4.0.1.jar                     slf4j-api-1.6.2.jar
jsr305-3.0.2.jar                                     opensaml-messaging-impl-4.0.1.jar                    transport-netty4-8.6.0-SNAPSHOT.jar
lang-tag-1.7.jar                                     opensaml-profile-api-4.0.1.jar                       x-pack-security-8.6.0-SNAPSHOT.jar
log4j-slf4j-impl-2.19.0.jar                          opensaml-profile-impl-4.0.1.jar                      xmlsec-2.1.4.jar
```

justincr-elastic commented:
@elastic/es-security, any ideas how to unblock the OIDC test fixture to apply the Security Manager policy?

I see my changes in /usr/share/elasticsearch/modules/x-pack-security/plugin-security.policy, but it does not appear to be read on the classpath.

justincr-elastic committed:
Merge branch 'citest/random-failures-89509-use-securerandom-with-seed' of github.com:justincr-elastic/elasticsearch into citest/random-failures-89509-use-securerandom-with-seed

justincr-elastic committed:
Add codeBase

justincr-elastic committed:
Merge branch 'main' of file:///home/q/github/mirrors/elasticsearch into citest/random-failures-89509-use-securerandom-with-seed

justincr-elastic committed:
Fix exception

justincr-elastic committed:
Use full jar names for codebases.

justincr-elastic committed:
Kludge: Remove codebases in test-framework.policy

justincr-elastic commented:
Depends on #91091

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into citest/random-failures-89509-use-securerandom-with-seed

justincr-elastic committed:
Remove secureRandom overrides.

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into citest/random-failures-89509-use-securerandom-with-seed

justincr-elastic commented:
@ChrisHegarty can you have a look at the updated policies? I changed to jar names, based on what I found when debugging PolicyUtil.java. When computing codebase names from jars, it only strips the version for pattern `name-#-#.jar` format. The two new jars use pattern `name-#-#-#.jar`, so I updated the codebase names in the policies.

The changes work in unit tests, but not the test fixture. I am not clear what is the difference between the unit tests, except maybe classes versus build jars. Is there a way to get extra diagnostics from SecurityManager to try and diagnose the problem?
